ompi/mca/coll/acoll/coll_acoll_utils.h

```diff
@@ -0,0 +1,769 @@
+    data->l1_gp_size = ompi_comm_size(subc->numa_comm);
+    data->l1_local_rank = ompi_comm_rank(subc->numa_comm);
+
+    comm_grp_ranks_local(comm, subc->numa_comm_ldrs, &tmp1, &tmp2, &data->l2_gp, tmp3);
+    data->l2_gp_size = ompi_comm_size(subc->numa_comm_ldrs);
+    data->l2_local_rank = ompi_comm_rank(subc->numa_comm_ldrs);
+    data->offset[0] = 16 * 1024;
+    data->offset[1] = data->offset[0] + size * 64;
+    data->offset[2] = data->offset[1] + size * 64;
+    data->offset[3] = data->offset[2] + rank * 8 * 1024;
+    data->allshmseg_id = (opal_shmem_ds_t *) malloc(sizeof(opal_shmem_ds_t) * size);
+    data->allshmmmap_sbuf = (void **) malloc(sizeof(void *) * size);
+    data->sync[0] = 0;
+    data->sync[1] = 0;
+    char *shfn;
+
+    /* Only the leaders need to allocate shared memory */
+    /* remaining ranks move their data into their leader's shm */
+    if (data->l1_gp[0] == rank) {
+        subc->initialized_shm_data = true;
+        ret = asprintf(&shfn, "/dev/shm/acoll_coll_shmem_seg.%u.%x.%d:%d-%d", geteuid(),
+                       OPAL_PROC_MY_NAME.jobid, ompi_comm_rank(MPI_COMM_WORLD),
+                       ompi_comm_get_local_cid(comm), ompi_comm_size(comm));
+    }
+
+    if (ret < 0) {
+        line = __LINE__;
+        goto error_hndl;
+    }
+
+    opal_shmem_ds_t seg_ds;
+    if (data->l1_gp[0] == rank) {
+        /* Assuming cacheline size is 64 */
+        long memsize
+            = (16 * 1024 /* scratch leader */ + 64 * size /* sync variables l1 group*/
+               + 64 * size /* sync variables l2 group*/ + 8 * 1024 * size /*data from ranks*/);
+        ret = opal_shmem_segment_create(&seg_ds, shfn, memsize);
+        free(shfn);
+    }
+
+    if (ret != OPAL_SUCCESS) {
+        opal_output_verbose(MCA_BASE_VERBOSE_ERROR, ompi_coll_base_framework.framework_output,
+                            "coll:acoll: Error: Could not create shared memory segment");
+        line = __LINE__;
+        goto error_hndl;
+    }
+
+    ret = comm->c_coll->coll_allgather(&seg_ds, sizeof(opal_shmem_ds_t), MPI_BYTE,
+                                       data->allshmseg_id, sizeof(opal_shmem_ds_t), MPI_BYTE, comm,
+                                       comm->c_coll->coll_allgather_module);
+
+    if (data->l1_gp[0] != rank) {
+        data->allshmmmap_sbuf[data->l1_gp[0]] = opal_shmem_segment_attach(
+            &data->allshmseg_id[data->l1_gp[0]]);
+    } else {
+        for (int i = 0; i < data->l2_gp_size; i++) {
+            data->allshmmmap_sbuf[data->l2_gp[i]] = opal_shmem_segment_attach(
+                &data->allshmseg_id[data->l2_gp[i]]);
+        }
+    }
+
+    int offset = 16 * 1024;
+    memset(((char *) data->allshmmmap_sbuf[data->l1_gp[0]]) + offset + 64 * rank, 0, 64);
+    if (data->l1_gp[0] == rank) {
+        memset(((char *) data->allshmmmap_sbuf[data->l2_gp[0]]) + (offset + 64 * size) + 64 * rank,
+               0, 64);
+    }
```

Comment:
Same here. I'd prefer to have names for numbers instead of magic numbers strewn across the code.

Reply:
Done